worldedit-core/src/main/java/com/fastasyncworldedit/core/queue/implementation/ApplyTask.java

```diff
@@ -0,0 +1,241 @@
+            ApplyTask<F> subtask = null;
+            int minRegionX = this.minChunkX >> this.shift;
+            int minRegionZ = this.minChunkZ >> this.shift;
+            int maxRegionX = this.maxChunkX >> this.shift;
+            int maxRegionZ = this.maxChunkZ >> this.shift;
+            // This task covers multiple regions. Create one subtask per region
+            for (int regionX = minRegionX; regionX <= maxRegionX; regionX++) {
+                for (int regionZ = minRegionZ; regionZ <= maxRegionZ; regionZ++) {
+                    if (ForkJoinTask.getSurplusQueuedTaskCount() > Settings.settings().QUEUE.PARALLEL_THREADS) {
+                        // assume we can do a bigger batch of work here - the other threads are busy for a while
+                        processRegion(regionX, regionZ, this.shift);
+                        continue;
+                    }
+                    if (this.shift == 0 && !this.commonState.region.containsChunk(regionX, regionZ)) {
+                        // if shift == 0, region coords are chunk coords
+                        continue; // chunks not intersecting with the region don't need a task
+                    }
+
+                    // creating more tasks will likely help parallelism as other threads aren't *that* busy
+                    subtask = new ApplyTask<>(
+                            this.commonState,
+                            subtask,
+                            regionX << this.shift,
+                            ((regionX + 1) << this.shift) - 1,
+                            regionZ << this.shift,
+                            ((regionZ + 1) << this.shift) - 1,
+                            this.shift
+                    );
+                    subtask.fork();
+                }
+            }
+            // try processing tasks in reverse order if not processed already, otherwise "wait" for completion
+            while (subtask != null) {
+                if (subtask.tryUnfork()) {
+                    subtask.invoke();
+                } else {
+                    subtask.join();
+                }
+                subtask = subtask.before;
+            }
+        } else {
+            // we reached a task for a single chunk, let's process it
+            processChunk(this.minChunkX, this.minChunkZ);
+        }
+        if (this.shift == INITIAL_REGION_SHIFT) {
+            onCompletion();
+        }
+    }
+
+    private void processRegion(int regionX, int regionZ, int shift) {
+        final ThreadState<F> state = getState();
+        this.commonState.parallelQueueExtent.enter(state.queue);
+        try {
+            for (int chunkX = regionX << shift; chunkX <= ((regionX  + 1) << shift) - 1; chunkX++) {
+                for (int chunkZ = regionZ << shift; chunkZ <= ((regionZ  + 1) << shift) - 1; chunkZ++) {
+                    if (!this.commonState.region.containsChunk(chunkX, chunkZ)) {
```

Comment:
Is it worth adding a micro-optimisation here to bound the initial and final chunkX|Z to a precomputed min/max? I suppose given it's not in the edit itself isn't necessarily an issue, but this could be a relatively complex operation depending on the region implementation

Reply:
It might help for certain region implementations, but I wanted to keep this change simple for now in case people encounter problems with it. I guess this could be a future improvement.